Pinta.Effects/Effects/TileEffect.cs

```diff
@@ -166,32 +172,65 @@ private static ColorBgra GetFinalPixelColor (
 			float finalV = settings.sin * transformedS + settings.cos * transformedT;
 
 			// Translate back to image coordinates
-			int unwrappedSampleX = (int) (settings.halfWidth + finalU);
-			int unwrappedSampleY = (int) (settings.halfHeight + finalV);
+			float preliminaryX = settings.halfWidth + finalU;
+			float preliminaryY = settings.halfHeight + finalV;
+
+			samples[p] = GetSample (
+				settings,
+				source,
+				sourceData,
+				original,
+				preliminaryX,
+				preliminaryY);
+		}
 
-			// Ensure coordinates wrap around the image dimensions
+		return ColorBgra.Blend (samples);
+	}
+
+	private ColorBgra GetSample (
+		TileSettings settings,
+		ImageSurface source,
+		ReadOnlySpan<ColorBgra> sourceData,
+		ColorBgra original,
+		float preliminaryX,
+		float preliminaryY)
+	{
+		if (IsOnSurface (settings, preliminaryX, preliminaryY))
+			return source.GetBilinearSample (preliminaryX, preliminaryY);
+
+		return settings.edgeBehavior switch {
+			EdgeBehavior.Clamp => source.GetBilinearSampleClamped (sourceData, settings.size.Width, settings.size.Height, preliminaryX, preliminaryY),
+			EdgeBehavior.Wrap => source.GetBilinearSampleWrapped (sourceData, settings.size.Width, settings.size.Height, preliminaryX, preliminaryY),
+			EdgeBehavior.Reflect => source.GetBilinearSampleClamped (sourceData, settings.size.Width, settings.size.Height, ReflectCoord (preliminaryX, settings.size.Width), ReflectCoord (preliminaryY, settings.size.Height)),
+			EdgeBehavior.Primary => palette.PrimaryColor.ToColorBgra (),
+			EdgeBehavior.Secondary => palette.SecondaryColor.ToColorBgra (),
+			EdgeBehavior.Transparent => ColorBgra.Transparent,
+			EdgeBehavior.Original => original,
+			_ => throw new ArgumentException ($"{nameof (settings.edgeBehavior)} is out of range", nameof (settings)),
+		};
+	}
 
-			int wrappedSampleX = (unwrappedSampleX + settings.size.Width) % settings.size.Width;
-			int wrappedSampleY = (unwrappedSampleY + settings.size.Height) % settings.size.Height;
+	private static bool IsOnSurface (TileSettings settings, float u, float v)
+		=> (u >= 0 && u <= (settings.size.Width - 1) && v >= 0 && v <= (settings.size.Height - 1));
 
-			int adjustedSampleX =
-				(wrappedSampleX < 0)
-				? (wrappedSampleX + settings.size.Width) % settings.size.Width
-				: wrappedSampleX;
-			int adjustedSampleY =
-				(wrappedSampleY < 0)
-				? (wrappedSampleY + settings.size.Height) % settings.size.Height
-				: wrappedSampleY;
+	private static float ReflectCoord (float value, int max)
```

Comment:
This looks the same as the one in `Warp.cs`, so perhaps it should go in some common location to be reused?
e.g maybe a `GetBilinearSampleReflected()` could make sense

Reply:
@cameronwhite I just added the extension method you suggested to `CairoExtensions`

Reply:
Thanks, that extension method looks good 👍